design/ip-address-management-for-networkdata.md

```diff
@@ -0,0 +1,373 @@
+      prefix: 24
+    - start: 192.168.1.10
+      end: 192.168.1.15
+      subnet: 192.168.1.0/24
+      gateway: 192.168.1.1
+      prefix: 24
+  gateway: 192.168.1.1
+  prefix: 24
+  allocations:
+    "metal3data-10": 192.168.0.9
+    "metal3data-9": 192.168.0.8
+  namePrefix: "provisioning"
+status:
+  lastUpdated: "2020-04-02T06:36:09Z"
+  addresses:
+    "192.168.0.11": "metal3data-1"
+  allocations:
+    "metal3data-1": "pool-1-192-168-0-11"
+```
+
+In the *spec* of *Metal3IPPool*, there would be a *pools* list, that would
+contain a list of IP address pools, with the *start* and *end* attributes
+giving the start and end ip addresses of the pool. The *subnet* field allows to
+verify that the allocated IP is in the pool and from which the start and end ip
+addresses can be inferred. Specifying single ip addresses can be achieved by
+setting the start and end ip address to that single ip address.
+
+The *prefix* and *gateway* parameters can be given for each pool of the list,
+or globally. If they are given for a pool they will override the global
+settings, that are default values. The *prefix* and *gateway* will be set on
+the *Metal3IPAddress* and can be fetched from a *Metal3DataTemplate*.
+
+The *allocations* fields is a map of object name and ip address that allow a
+user to specify a set of static allocations for some objects.
+
+The *namePrefix* contains the prefix used to name the Metal3IPAddress objects
+created. It must remain the same for a subnet, across updates or changes in the
+Metal3IPPool object to keep the existing leases.
+
+The *status* would contain a *lastUpdated* field with the timestamp of the last
+update. In case of an error during the allocation (pool exhaustion for example),
+the error would be reported on the Consumer object, the *error* boolean and
+*errorMessage* field on the *Metal3Data* object.
+The *allocations* map will map the IP address to the *Metal3Data* object it was
+allocated for and the *addresses* will map the *metal3Data* objects with the
+*Metal3IPAddress* objects.
+
+The *Metal3IPAddress* object would be the following
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3IPAddress
+metadata:
+  name: pool-1-192-168-0-11
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    kind: Metal3IPPool
+    name: pool-1
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    kind: Metal3Data
+    name: metal3data-1
+spec:
+  Owner:
+    Name: metal3data-1
+  Address: 192.168.0.11
+  prefix: 24
+  gateway: 192.168.0.1
+  Metal3IPPool:
+    Name: pool-1
+status:
+  ready: true
+```
+
+For each owner reference added on a *Metal3IPPool* that is not a *Metal3IPPool*,
+the controller reconciling the *Metal3IPPool* will select an available IP
+address randomly from the available IP addresses, if the object name is not in
+the *allocations* map in the object *spec* or in the *addresses* map in the
+*status*.
```

Comment:
What do you think about using a claim-based API instead of using owner references? Something similar to PersistentVolumeClaim and PersistentVolume, for example. Or the way we treat Machine and BareMetalHost, even.

That would give us a place to record state for each request for an IP in a consistent way. It would also be more familiar to users of other APIs.

An IPAddressClaim could replace the Metal3Address described here. It would need a spec field with a reference to the pool to be used (maybe later it could also have some less direct way to pick a pool, like a label selector) and it would have status fields for the results of the allocation. Alternatively, the claim could be a 3rd CRD and could just link to an address CR that holds the results.

Reply:
Good point. It would be probably more understandable. However, I would still keep the Metal3IPAddress separately as that allows to prevent duplicates of allocations based on the naming scheme (two Metal3IPAddress with the same address would have the same name). That prevents conflict in a very straightforward way. Otherwise we would have issues making sure the states are synced across different objects, so I'd definitely keep it separated. About making it a 3rd CRD, this is possible, and I like the idea. 

The only doubt I would have is that we would multiply the number of objects by 2 for the IPAddresses. We could also apply the same idea for the Data template and Data. But it probably would be fine (100 nodes -> 400 addresses, datas, and claim objects for example).